toolkit/offchain/src/plutus_script.rs

```diff
@@ -0,0 +1,68 @@
+use anyhow::{anyhow, Context};
+use cardano_serialization_lib::{Address, LanguageKind, NetworkIdKind, PlutusData};
+use plutus::ToDatum;
+use sidechain_domain::{MainchainAddressHash, PolicyId};
+use uplc::ast::{DeBruijn, Program};
+
+use crate::{csl::*, untyped_plutus::*};
+
+/// Wraps a Plutus script cbor
+pub struct PlutusScript {
+	pub bytes: Vec<u8>,
+	pub language: LanguageKind,
+}
+
+impl PlutusScript {
+	pub fn from_cbor(cbor: &[u8], language: LanguageKind) -> Self {
+		Self { bytes: cbor.into(), language }
+	}
+	pub fn from_wrapped_cbor(cbor: &[u8], language: LanguageKind) -> anyhow::Result<Self> {
+		Ok(Self::from_cbor(&unwrap_one_layer_of_cbor(cbor)?, language))
+	}
+
+	pub fn apply_data(self, data: impl ToDatum) -> Result<Self, anyhow::Error> {
+		let data = datum_to_uplc_plutus_data(&data.to_datum());
+		self.apply_uplc_data(data)
+	}
+
+	pub fn apply_uplc_data(self, data: uplc::PlutusData) -> Result<Self, anyhow::Error> {
+		let mut buffer = Vec::new();
+		let mut program = Program::<DeBruijn>::from_cbor(&self.bytes, &mut buffer)
+			.map_err(|e| anyhow!(e.to_string()))?;
+		program = program.apply_data(data);
+		let bytes = program
+			.to_cbor()
+			.map_err(|_| anyhow!("Couldn't encode resulting script as CBOR."))?;
+		Ok(Self { bytes: bytes.into(), ..self })
+	}
+
+	/// Builds an CSL `Address` for plutus script from the data obtained from smart contracts.
+	pub fn plutus_address(&self, network: NetworkIdKind) -> Address {
```

Comment:
Regarding names of this and following functions, I think that `plutus_` prefix is redundant.